src/restore.ts

```diff
@@ -28,7 +29,9 @@ async function run() {
         // pre-clean the target directory on cache mismatch
         const packages = await getPackages();
```

Comment:
the important thing here would be to make `getPackages` aware of the different dirs. `target-dir` may even be the wrong thing here, and it should rather be `manifest-dir`, since this collects a set of packages that we do need to keep.

Reply:
Ok I think I'm getting a picture of how this works. So `getPackages` is intended to collect a list of packages that do not belong to the current project, i.e. packages that can be cached, right? And it does this by checking `cargo metadata` and finding all packages that do not share a path prefix with `cwd`.

So what you're saying is that perhaps we should have a list of project roots where we will find `Cargo.toml`'s, which you are calling `manifest-dir`, and we should repeat this process for each of those project roots, right?

Reply:
Yes, your analysis is spot on.
I guess `manifest-path` is the right term for it rather: https://doc.rust-lang.org/cargo/commands/cargo-test.html#manifest-options

Either way, I think just merging all the dependencies from all the projects should be fine.

Reply:
@Swatinem what do you think of `workspace-paths` instead? Because I believe that's actually what we need to determine. If you run `cargo metadata` from a workspace directory, it will include the `manifest_path` of each package in the workspace. Plus that is more aligned with the original "target-dir" idea, since each unique workspace would indeed have it's own target directory.